service/qos/service_level_controller.cc

```diff
@@ -317,39 +334,45 @@ void service_level_controller::update_from_distributed_data(std::function<steady
     }
 }
 
-future<> service_level_controller::add_distributed_service_level(sstring name, service_level_options slo, bool if_not_exists) {
+future<> service_level_controller::add_distributed_service_level(sstring name, service_level_options slo, bool if_not_exists, std::optional<service::group0_guard> guard) {
     set_service_level_op_type add_type = if_not_exists ? set_service_level_op_type::add_if_not_exists : set_service_level_op_type::add;
-    return set_distributed_service_level(name, slo, add_type);
+    return set_distributed_service_level(name, slo, add_type, std::move(guard));
 }
 
-future<> service_level_controller::alter_distributed_service_level(sstring name, service_level_options slo) {
-    return set_distributed_service_level(name, slo, set_service_level_op_type::alter);
+future<> service_level_controller::alter_distributed_service_level(sstring name, service_level_options slo, std::optional<service::group0_guard> guard) {
+    return set_distributed_service_level(name, slo, set_service_level_op_type::alter, std::move(guard));
 }
 
-future<> service_level_controller::drop_distributed_service_level(sstring name, bool if_exists) {
-    return _sl_data_accessor->get_service_levels().then([this, name, if_exists] (qos::service_levels_info sl_info) {
-        auto it = sl_info.find(name);
-        if (it == sl_info.end()) {
-            if (if_exists) {
-                return make_ready_future();
-            } else {
-                return make_exception_future(nonexistant_service_level_exception(name));
-            }
+future<> service_level_controller::drop_distributed_service_level(sstring name, bool if_exists, std::optional<service::group0_guard> guard) {
+    auto sl_info = co_await _sl_data_accessor->get_service_levels();
+    auto it = sl_info.find(name);
+    if (it == sl_info.end()) {
+        if (if_exists) {
+            co_return;
         } else {
-            auto& role_manager = _auth_service.local().underlying_role_manager();
-            return role_manager.query_attribute_for_all("service_level").then( [&role_manager, name] (auth::role_manager::attribute_vals attributes) {
-                return parallel_for_each(attributes.begin(), attributes.end(), [&role_manager, name] (auto&& attr) {
-                    if (attr.second == name) {
-                        return do_with(attr.first, [&role_manager] (const sstring& role_name) {return role_manager.remove_attribute(role_name, "service_level");});
-                    } else {
-                        return make_ready_future();
-                    }
-                });
-            }).then([this, name] {
-                return _sl_data_accessor->drop_service_level(name);
-            });
+            throw nonexistant_service_level_exception(name);
         }
-    });
+    } else {
```

Comment:
nit: no need for else, the code above either returns or throws so code below doesn't need to be nested